File: baselibrary/src/main/java/com/aslan/baselibrary/http/response/ResponseFunctionFlowable.java
package com.aslan.baselibrary.http.response;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.aslan.baselibrary.base.DataError;
import io.reactivex.Flowable;

/**
 * 处理业务异常
 *
 * @author Aslan chenhengfei@yy.com
 * @date 2020/6/12
 */
public class ResponseFunctionFlowable<T> extends BaseResponseFunction<T, Flowable<T>> {

  public ResponseFunctionFlowable(Context context) {
    super(context);
  }

  @Override
  public Flowable<T> error(@NonNull DataError ex) {
    return Flowable.error(ex);
  }

  @Override
  public Flowable<T> handleData(@Nullable T item) {
    if (item == null) {
      return Flowable.error(new NullPointerException("respone data is empty"));
    } else {
      return Flowable.just(item);
    }
  }
}

File: baselibrary/src/main/java/com/aslan/baselibrary/http/response/ResponseFunctionFlowable.kt
package com.aslan.baselibrary.http.response

import android.content.Context
import com.aslan.baselibrary.R
import io.reactivex.Flowable

/**
 * 处理业务异常
 *
 * @author Aslan chenhengfei@yy.com
 * @date 2020/6/12
 */
class ResponseFunctionFlowable<T>(val context: Context) : BaseResponseFunction<T, Flowable<T>>(context) {
    override fun error(ex: Exception): Flowable<T> {
        return Flowable.error(ex)
    }

    override fun handleData(item: T?): Flowable<T> {
        return if (item == null) {
            Flowable.error(NullPointerException(context.getString(R.string.error_data_empty)))
        } else {
            Flowable.just(item)
        }
    }
}